### PR TITLE
ENH: Add a `mamba` plugin

### LINKED ISSUE #962: Ability to use mamba instead of conda for environment creation
For more complex environments (eg in the pandas asv suite), the step of creating the environment can take quite a long time with conda. Using mamba could speed up this step.

### PR DESCRIPTION
**Pull request by HaoZeke**
This is much faster than the `conda` solver and is more robust with a better Pythonic API. Closes #962.

### CODE REVIEW

**Comment by HaoZeke** [comment]
Expects #1237

**Comment by zacharyburnett** [comment]
@HaoZeke nice PR, this should speed up the benchmarking process! You can also try using the https://github.com/marketplace/actions/provision-with-micromamba action in the CI to make it even faster

**Comment by HaoZeke** [comment]
> @HaoZeke nice PR, this should speed up the benchmarking process! You can also try using the https://github.com/marketplace/actions/provision-with-micromamba action to make the CI even faster

Yup, however, I think I'd need to add another plugin for `micromamba` since it isn't covered by the `python` API used here.

**Comment by HaoZeke** [comment]
Installing dependencies on `pypy-3.7` leads to an obscure error:

```
Installing collected packages: sphinx_bootstrap_theme, sortedcontainers, snowballstemmer, sgmllib3k, pytz, python-hglib, mccabe, distlib, charset-normalizer, zipp, urllib3, typing-extensions, tomli, sphinxcontrib-serializinghtml, sphinxcontrib-qthelp, sphinxcontrib-jsmath, sphinxcontrib-htmlhelp, sphinxcontrib-devhelp, sphinxcontrib-applehelp, sniffio, pyyaml, PySocks, Pygments, pyflakes, pycodestyle, packaging, numpy, MarkupSafe, isort, iniconfig, imagesize, idna, filelock, feedparser, execnet, exceptiongroup, docutils, certifi, babel, attrs, async-generator, alabaster, requests, platformdirs, outcome, Jinja2, importlib-metadata, h11, wsproto, virtualenv, trio, sphinx, pluggy, flake8, trio-websocket, pytest, selenium, pytest-xdist, pytest-timeout, pytest-rerunfailures
ERROR: Could not install packages due to an OSError: [Errno 2] No such file or directory: '/usr/share/miniconda/envs/test/site-packages/../bin/rst2html.py'

Error: Process completed with exit code 1.
```

Dropping support for `3.7` in general might be a good idea? Especially since newer plugins (e.g. `mamba`, `memray`) require newer versions as well.

File: asv/environment.py [hunk 1]
```diff
@@ -428,18 +438,24 @@ def get_environment_class(conf, python):
         if cls.matches_python_fallback and cls.matches(python):
             return cls
     raise EnvironmentUnavailable(
-        "No way to create environment for python='{0}'".format(python))
+        "No way to create environment for python='{0}'".format(python)
+    )
```
Comment (mattip) [inline asv/environment.py:442]: f-strings might be nicer

File: asv/environment.py [hunk 2]
```diff
@@ -428,18 +438,24 @@ def get_environment_class(conf, python):
...
 
 def get_environment_class_by_name(environment_type):
     """
     Find the environment class with the given name.
     """
     for cls in util.iter_subclasses(Environment):
+        if "mamba" in environment_type and (
+            sys.version_info.major >= 3 and sys.version_info.minor < 8
+        ):
+            environment_type = "conda"
```
Comment (mattip) [inline asv/environment.py:453]: why the upper limit? Also: what happens if "conda" is used instead of "mamba"?

**Review by mattip — COMMENTED** [review]
It is much easier to review if the formatting changes are done in a separate PR, and not mixed in with code changes.

Rather than hard-code `pip` to install the project, shouldn't there be a way for the project to say what backend to use?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -23,20 +23,26 @@ jobs:
         with:
           python-version: 3.7
 
-      - name: Set up Python ${{ matrix.python-version }}
-        uses: actions/setup-python@v4
+      - name: Setup Conda with ${{ matrix.python-version }}
+        uses: conda-incubator/setup-miniconda@v2
```
Comment (fangchenli) [inline .github/workflows/ci.yml:27]: Setting up conda and memba in the CI takes longer than using `setup-python`. For a library with very few dependencies, it's not worth it. Also, the CI changes are unrelated to the topic of this PR and should be addressed separately.

**Review by fangchenli — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 3 above)
Comment (HaoZeke) [reply .github/workflows/ci.yml:27]: Makes sense, originally I thought of testing the new environment handler in CI as well, which is why I wanted to add it here, but I guess that could be elsewhere.

**Comment by HaoZeke** [comment]
Some benchmarks for cold starts `time asv run` with `.asv` removed:

```bash
# Conda
asv run  63.61s user 4.06s system 101% cpu 1:06.50 total
# Mamba
asv run  13.42s user 2.43s system 120% cpu 13.152 total
# Virtualenv
asv run  8.35s user 1.83s system 116% cpu 8.751 total
```

For this [toy project](https://github.com:HaoZeke/asv_experiments)

**Comment by HaoZeke** [comment]
> It is much easier to review if the formatting changes are done in a separate PR, and not mixed in with code changes.
Sorry, reverted those.

> Rather than hard-code `pip` to install the project, shouldn't there be a way for the project to say what backend to use?

The project is installed with the `install_command` configuration option, the plugin only uses `pip` for the dependencies..

File: asv/environment.py (hunk 2 above)
Comment (HaoZeke) [reply asv/environment.py]: If `conda` is used it continues to use the old solver. The idea here was that it should silently fall back to the `conda` mechanism before `libmambapy` is present. The upper limit shouldn't be there, sorry.

**Comment by HaoZeke** [comment]
I have cleaned this up and simplified it. There are no tests, other than manual ones. Tests would require changes to the CI which can be done elsewhere.

File: asv/environment.py [hunk 4]
```diff
@@ -428,7 +428,7 @@ def get_environment_class(conf, python):
         if cls.matches_python_fallback and cls.matches(python):
             return cls
     raise EnvironmentUnavailable(
-        "No way to create environment for python='{0}'".format(python))
+        "No way to create environment for python='{python}'")
```
Comment (zacharyburnett) [inline asv/environment.py:431]: does this need an `f`? You could use `flynt` or something to do these automatically (maybe out of scope for this PR)

Comment (HaoZeke) [reply asv/environment.py:431]: Yup, good eye. Also I'll open another PR for `flynt` but I think @jarrodmillman has converted a lot of it already.

**Comment by mattip** [comment]
Thanks @HaoZeke

**Pull request merged**